File: Temporal_Tree_tests/ApproccioRicorsivo-BottomUp/Approccio2-FUNZIONANTE/test/DFS-EA-Tmax.py
from timeit import default_timer as timer
from datetime import timedelta

class Node:
    def __init__(self, value, weight=[]):
        self.value = value
        self.weight = weight
        self.left = None
        self.right = None


def binary_search(arr, target):
    if len(arr) == 1:  # Caso in cui l'array ha un solo elemento
        return arr[0] if arr[0] >= target else -1  # Restituisce l'indice se il valore è >= target
    left, right = 0, len(arr) - 1
    result = -1  # Inizialmente, supponiamo che non ci sia un valore valido

    while left <= right:
        mid = (left + right) // 2
        
        if arr[mid] >= target:
            result = mid  # Salviamo l'indice come potenziale risultato
            right = mid - 1  # Continuiamo a cercare nella metà sinistra
        else:
            left = mid + 1  # Cerchiamo nella metà destra
    if result == -1:
        return -1
    else:
        return arr[result]
    
def binary_search_leq(arr, target):
    if len(arr) == 1:  # Caso in cui l'array ha un solo elemento
        return arr[0] if arr[0] <= target else -1  # Restituisce l'elemento se è ≤ target
    
    left, right = 0, len(arr) - 1
    result = -1  # Inizialmente, supponiamo che non ci sia un valore valido

    while left <= right:
        mid = (left + right) // 2
        
        if arr[mid] <= target:
            result = mid  # Salviamo l'indice come potenziale risultato
            left = mid + 1  # Continuiamo a cercare nella metà destra
        else:
            right = mid - 1  # Cerchiamo nella metà sinistra

    if result == -1:
        return -1  # Nessun valore trovato ≤ target
    else:
        return arr[result]

def dfs_EA_tmax_spazio1(root):
    
    if root is None:
        return float("-inf"),float("inf")
    if root.left == None and root.right == None:
        print(f"EA e tempo max visita per il sottoalbero radicato nel nodo {root.value} (foglia) : {root.weight[0],root.weight[-1]}")
        return root.weight[0],root.weight[-1] #min(root.weight), max(root.weight)

    min_sx,max_sx = dfs_EA_tmax_spazio1(root.left)

    min_dx,max_dx = dfs_EA_tmax_spazio1(root.right)
    
    if min_sx>max_dx or min_dx>max_sx:
        return float("inf"),float("inf")
    
    EA = max(min_sx,min_dx)
    t_max_visita = min(max_sx,max_dx)
    print(f"EA e tempo max visita per il sottoalbero radicato nel nodo {root.value} (nodo interno) : {EA,t_max_visita}")
    k = binary_search(root.weight,EA)
    nextTimeMax = binary_search_leq(root.weight,t_max_visita) #binary search per trovare il predecessore, quindi il primo tempo t <= t_max_visita
    if k == -1 or nextTimeMax == -1 or nextTimeMax < t_max_visita:
        return float("inf"),float("inf")
        #exit("Errore: EA o tempo max visita non trovati")
    minTime = min(t_max_visita,nextTimeMax)
    #valori_EA_Tmax=(EA,t_max_visita)

    return k,minTime

def dfs_EA_tmax_spazioN(root):
    # Caso base: nodo nullo
    if root is None:
        return {}

    sottoalberi = {}
    # Caso base: foglia
    if root.left is None and root.right is None:
        print(f"EA e tempo max visita per il sottoalbero radicato nel nodo {root.value} (foglia): {root.weight[0], root.weight[-1]}")
        #return {root.value: (root.weight[0], root.weight[-1])}
        sottoalberi[root.value] = (root.weight[0], root.weight[-1])
        return sottoalberi
    # Variabili per raccogliere i valori EA e Tmax per ogni sottoalbero
    

    # Calcolo ricorsivo per il sottoalbero sinistro
    if root.left is not None:
        sottoalberi.update(dfs_EA_tmax_spazioN(root.left))

    # Calcolo ricorsivo per il sottoalbero destro
    if root.right is not None:
        sottoalberi.update(dfs_EA_tmax_spazioN(root.right))

    # Estrai i valori di EA e Tmax dai figli
    ea_sx, t_max_sx = sottoalberi[root.left.value] if root.left else (float("-inf"), float("inf"))
    ea_dx, t_max_dx = sottoalberi[root.right.value] if root.right else (float("-inf"), float("inf"))

    # Controllo di consistenza tra i sottoalberi
    if ea_sx > t_max_dx or ea_dx > t_max_sx:
        #return {root.value: (float("inf"), float("inf"))}
        sottoalberi[root.value] = (float("inf"), float("inf"))
        return sottoalberi

    # Calcolo EA e Tmax per il nodo corrente
    EA = max(ea_sx, ea_dx)
    t_max_visita = min(t_max_sx, t_max_dx)
    print(f"EA e tempo max visita per il sottoalbero radicato nel nodo {root.value} (nodo interno): {EA, t_max_visita}")
    
    k = binary_search(root.weight,EA)
    nextTimeMax = binary_search_leq(root.weight,t_max_visita) #binary search per trovare il predecessore, quindi il primo tempo t <= t_max_visita
    if root.weight == []:
        k,nextTimeMax = 0,0
        sottoalberi[root.value] = (k, nextTimeMax)
        return sottoalberi
    if k == -1 or nextTimeMax == -1 or nextTimeMax < t_max_visita:
        sottoalberi[root.value] = (float("inf"), float("inf"))
        #return {root.value: (float("inf"), float("inf"))}
        return sottoalberi
    minTime = min(t_max_visita,nextTimeMax)
    # Aggiornamento del nodo corrente nei risultati
    sottoalberi[root.value] = (k, minTime)

    return sottoalberi


def algoritmo(root):
    print("Questo è per alberi binari\n")
    
    ea_sx,t_max_sx = dfs_EA_tmax_spazio1(root.left)
    ea_dx,t_max_dx = dfs_EA_tmax_spazio1(root.right)
    #print(valori_sx,valori_dx)
    print("------------------------------------------------")
    print(f"EA e tempo max visita sx della radice {root.value} : {ea_sx,t_max_sx}")
    print(f"EA e tempo max visita dx della radice {root.value} : {ea_dx,t_max_dx}")
    
    if ea_sx == float("inf") or ea_dx == float("inf"):
       return False
    
    # Ogni controllo del caso per alberi non binari
    if ea_sx <= t_max_dx and ea_dx <= t_max_sx:
       return True
    else:
       return False
    
def algoritmo2(root):
    print("Questo è per alberi binari\n")
    
    # Esegui DFS-EA-Tmax una sola volta
    risultati = dfs_EA_tmax_spazioN(root)

    # Ottieni i risultati per i figli della radice
    ea_sx, t_max_sx = risultati[root.left.value] if root.left else (float("-inf"), float("inf"))
    ea_dx, t_max_dx = risultati[root.right.value] if root.right else (float("-inf"), float("inf"))

    print("------------------------------------------------")
    print(f"EA e tempo max visita sx della radice {root.value}: {ea_sx, t_max_sx}")
    print(f"EA e tempo max visita dx della radice {root.value}: {ea_dx, t_max_dx}")

    # Controllo finale per la radice
    if ea_sx == float("inf") or ea_dx == float("inf"):
        return False

    if ea_sx <= t_max_dx and ea_dx <= t_max_sx:
        return True
    else:
        return False

def print_tree(root, level=0):
    if root is not None:
        print("  " * level + f"Node {root.value}, Weight: {root.weight}")
        for child in root.children:
            print_tree(child, level + 1)


# Esempio 1
# root = Node('A')
# root.left = Node('B', weight=[3])
# root.right = Node('C', weight=[1,4])
# root.left.left = Node('D', weight=[1,3])
# root.left.right = Node('E', weight=[2,5])

# Esempio 2
root = Node('A')
root.left = Node('B', weight=[2,6])
root.right = Node('C', weight=[6])
root.left.left = Node('D', weight=[1,2,3,4,5])
root.right.right = Node('E', weight=[6])

# Esempio 3
# root = Node('A')
# root.left = Node('B', weight=[3])
# root.right = Node('C', weight=[2, 5])
# root.right.right = Node('D', weight=[1,2,3,4,5,6,7])

# Esempio 4
# root = Node('A')
# root.left = Node('B', weight=[2])
# root.right = Node('C', weight=[5,7])
# root.left.left = Node('D', weight=[5])
# root.right.left = Node('E', weight=[8,9])
# root.right.right = Node('F', weight=[1,2])

# Esempio 5
# root = Node('A')
# root.left = Node('B', weight=[1, 3,4])
# root.right = Node('C', weight=[4])
# root.left.left = Node('D', weight=[2,4])
# root.right.right = Node('E', weight=[2,5,6])

# Esempio 6
# root = Node('A')
# root.left = Node('B', weight=[1, 3])
# root.right = Node('C', weight=[2, 6])
# root.left.left = Node('D', weight=[3,5])
# root.left.right = Node('E', weight=[4,6])

# Esempio 7
# root = Node('A',weight=[float("-inf")])
# root.left = Node('B', weight=[1,3])
# root.left.left = Node('D', weight=[1,2])
# root.left.right = Node('E', weight=[2,4])

# Esempio 8
# root = Node('A')
# root.left = Node('B', weight=[1,4])
# root.left.left = Node('C', weight=[2,4])
# root.left.left.left = Node('D', weight=[1,3])
# root.left.left.left.left = Node('E', weight=[3,4])

# Esempio 9
# root = Node('A')
# root.left = Node('B', weight=[2,3,5,8])
# root.right = Node('C', weight=[3,5])
# root.left.left = Node('D', weight=[1,4,5,6])
# root.left.right = Node('E', weight=[1,5,9])
# root.right.left = Node('F', weight=[1,2,5])
# root.right.right = Node('G', weight=[2,5])
# root.left.left.left = Node('H', weight=[2,4])
# root.right.left.right = Node('I', weight=[1,11])

# Esempio 10
# root = Node('A')
# root.left = Node('B', weight=[1,9])
# root.right = Node('C', weight=[8])
# root.left.left = Node('D', weight=[2])
# root.left.right = Node('E', weight=[7])
# root.right.left = Node('F', weight=[4])
# root.right.right = Node('G', weight=[6])
# root.left.left.left = Node('H', weight=[3])
# root.left.left.right = Node('I', weight=[1,2,3])

# Esempio 11
# root = Node('A',root=True)
# root.left = Node('B', weight=[1,5,7])
# root.right = Node('C', weight=[2,5,8])
# root.left.left = Node('D', weight=[2,8])
# root.left.right = Node('E', weight=[2,5,10])
# root.left.right.left = Node('F', weight=[1,5,6])
# root.left.right.right = Node('G', weight=[2,7,9])
# root.right.right = Node('H', weight=[2,3,4])
# root.right.right.left = Node('I', weight=[1,15])
# root.right.right.right = Node('J', weight=[1,11])

start = timer()
print("Versione con spazio O(1)")
print(f"Albero temporalmente connesso? : {algoritmo(root)}")
print("\nVersione con spazio O(n)")
print(f"Albero temporalmente connesso? : {algoritmo2(root)}")
end = timer()
print(timedelta(seconds=end-start))

File: Temporal_Tree/test/DFS-EA-Tmax.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include <climits>
#include <cstdlib>
#include <chrono>
using namespace std::chrono;

struct Node {
    char value;
    std::vector<int> weight;
    Node* left;
    Node* right;

    Node(char val, std::vector<int> wt = {}) : value(val), weight(wt), left(nullptr), right(nullptr) {}
};

int binary_search(const std::vector<int>& arr, int target) {
    if (arr.empty()) return -1; // Verifica array vuoto
    int left = 0, right = arr.size() - 1;
    int result = -1;

    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (arr[mid] >= target) {
            result = mid;
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }
    return result != -1 ? arr[result] : -1;
}

int binary_search_leq(const std::vector<int>& arr, int target) {
    if (arr.empty()) return -1; // Verifica array vuoto
    int left = 0, right = arr.size() - 1;
    int result = -1;

    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (arr[mid] <= target) {
            result = mid;
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return result != -1 ? arr[result] : -1;
}

std::pair<int, int> dfs_EA_tmax_spazio1(Node* root) {
    if (!root) return {INT_MIN, INT_MAX};
    if (!root->left && !root->right) {
        if (root->weight.empty()) {
            std::cerr << "Errore: Nodo " << root->value << " ha una lista di pesi vuota!\n";
            exit(EXIT_FAILURE);
        }
        std::cout << "EA e tempo max visita per il sottoalbero radicato nel nodo " << root->value << " (foglia): (" << root->weight.front() << ", " << root->weight.back() << ")\n";
        return {root->weight.front(), root->weight.back()};
    }

    auto [min_sx, max_sx] = root->left ? dfs_EA_tmax_spazio1(root->left) : std::make_pair(INT_MIN, INT_MAX);
    auto [min_dx, max_dx] = root->right ? dfs_EA_tmax_spazio1(root->right) : std::make_pair(INT_MIN, INT_MAX);

    int EA = std::max(min_sx, min_dx);
    int t_max_visita = std::min(max_sx, max_dx);
    std::cout << "EA e tempo max visita per il sottoalbero radicato nel nodo " << root->value << " (nodo interno): (" << EA << ", " << t_max_visita << ")\n";

    int k = binary_search(root->weight, EA);
    int nextTimeMax = binary_search_leq(root->weight, t_max_visita);

    if (k == -1 || nextTimeMax == -1) {
        std::cerr << "Errore: EA o tempo max visita non trovati per il nodo " << root->value << "\n";
        exit(EXIT_FAILURE);
    }

    int minTime = std::min(t_max_visita, nextTimeMax);
    return {k, minTime};
}

std::map<char, std::pair<int, int>> dfs_EA_tmax_spazioN(Node* root) {
    if (!root) return {};

    if (!root->left && !root->right) {
        if (root->weight.empty()) {
            std::cerr << "Errore: Nodo " << root->value << " ha una lista di pesi vuota!\n";
            exit(EXIT_FAILURE);
        }
        std::cout << "EA e tempo max visita per il sottoalbero radicato nel nodo " << root->value << " (foglia): (" << root->weight.front() << ", " << root->weight.back() << ")\n";
        return {{root->value, {root->weight.front(), root->weight.back()}}};
    }

    std::map<char, std::pair<int, int>> sottoalberi;

    if (root->left) {
        auto left_result = dfs_EA_tmax_spazioN(root->left);
        sottoalberi.insert(left_result.begin(), left_result.end());
    }

    if (root->right) {
        auto right_result = dfs_EA_tmax_spazioN(root->right);
        sottoalberi.insert(right_result.begin(), right_result.end());
    }

    int ea_sx = root->left ? sottoalberi[root->left->value].first : INT_MIN;
    int t_max_sx = root->left ? sottoalberi[root->left->value].second : INT_MAX;
    int ea_dx = root->right ? sottoalberi[root->right->value].first : INT_MIN;
    int t_max_dx = root->right ? sottoalberi[root->right->value].second : INT_MAX;

    int EA = std::max(ea_sx, ea_dx);
    int t_max_visita = std::min(t_max_sx, t_max_dx);
    std::cout << "EA e tempo max visita per il sottoalbero radicato nel nodo " << root->value << " (nodo interno): (" << EA << ", " << t_max_visita << ")\n";

    int k = binary_search(root->weight, EA);
    int nextTimeMax = binary_search_leq(root->weight, t_max_visita);

    int minTime = std::min(t_max_visita, nextTimeMax);
    sottoalberi[root->value] = {k, minTime};

    return sottoalberi;
}

bool algoritmo(Node* root) {
    std::cout << "Versione con spazio O(1)\n";

    auto [ea_sx, t_max_sx] = root->left ? dfs_EA_tmax_spazio1(root->left) : std::make_pair(INT_MIN, INT_MAX);
    auto [ea_dx, t_max_dx] = root->right ? dfs_EA_tmax_spazio1(root->right) : std::make_pair(INT_MIN, INT_MAX);

    std::cout << "------------------------------------------------\n";
    std::cout << "EA e tempo max visita sx della radice " << root->value << ": (" << ea_sx << ", " << t_max_sx << ")\n";
    std::cout << "EA e tempo max visita dx della radice " << root->value << ": (" << ea_dx << ", " << t_max_dx << ")\n";

    return ea_sx <= t_max_dx && ea_dx <= t_max_sx;
}

bool algoritmo2(Node* root) {
    std::cout << "Versione con spazio O(n)\n";

    auto risultati = dfs_EA_tmax_spazioN(root);

    int ea_sx = root->left ? risultati[root->left->value].first : INT_MIN;
    int t_max_sx = root->left ? risultati[root->left->value].second : INT_MAX;
    int ea_dx = root->right ? risultati[root->right->value].first : INT_MIN;
    int t_max_dx = root->right ? risultati[root->right->value].second : INT_MAX;

    std::cout << "------------------------------------------------\n";
    std::cout << "EA e tempo max visita sx della radice " << root->value << ": (" << ea_sx << ", " << t_max_sx << ")\n";
    std::cout << "EA e tempo max visita dx della radice " << root->value << ": (" << ea_dx << ", " << t_max_dx << ")\n";

    return ea_sx <= t_max_dx && ea_dx <= t_max_sx;
}

int main() {
    // Node* root = new Node('A');
    // root->left = new Node('B', {2, 6});
    // root->right = new Node('C', {6});
    // root->left->left = new Node('D', {1, 2, 3, 4, 5,6});
    // root->right->right = new Node('E', {6});
    Node* root = new Node('A');
    root->left = new Node('B', {2, 3, 5, 8});
    root->right = new Node('C', {3, 5});
    root->left->left = new Node('D', {1, 4, 5, 6});
    root->left->right = new Node('E', {1, 5, 9});
    root->right->left = new Node('F', {1,2,5});
    root->right->right = new Node('G', {2, 5});
    root->left->left->left = new Node('H', {2, 4});
    root->right->left->right = new Node('I', {1, 11});

    auto start = high_resolution_clock::now();
    std::cout << "Albero temporalmente connesso? : " << (algoritmo(root) ? "Risposta : Albero Temporalmente Connesso" : "Fatte da Nculo") << "\n";
    std::cout << "\n";
    std::cout << "Albero temporalmente connesso? : " << (algoritmo2(root) ? "Risposta : Albero Temporalmente Connesso" : "Fatte da Nculo") << "\n";
    auto stop = high_resolution_clock::now();

    auto duration = duration_cast<microseconds>(stop - start);

    std::cout << "Time taken by function: "
         << duration.count() << " microseconds";
    return 0;
}